gitlab/resource_gitlab_project_approval_rule.go

```diff
@@ -0,0 +1,425 @@
+package gitlab
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+// https://docs.gitlab.com/ee/api/merge_request_approvals.html#create-project-level-rule
+
+var resourceGitLabProjectApprovalRuleName = "gitlab_project_approval_rule"
+
+var resourceGitLabProjectApprovalRuleSchema = map[string]*schema.Schema{
+	"project": {
+		Type:     schema.TypeString,
+		ForceNew: true,
+		Required: true,
+	},
+	"name": {
+		Type:     schema.TypeString,
+		ForceNew: true,
+		Required: true,
+	},
+	"approvals_required": {
+		Type:     schema.TypeInt,
+		Required: true,
+	},
+	"user_ids": {
+		Type:     schema.TypeSet,
+		Optional: true,
+		Elem:     &schema.Schema{Type: schema.TypeInt},
+		Set:      schema.HashInt,
+	},
+	"group_ids": {
+		Type:     schema.TypeSet,
+		Optional: true,
+		Elem:     &schema.Schema{Type: schema.TypeInt},
+		Set:      schema.HashInt,
+	},
+	"eligible_approvers": {
+		Type:     schema.TypeSet,
+		Computed: true,
+		Elem:     &schema.Schema{Type: schema.TypeInt},
+		Set:      schema.HashInt,
+	},
+}
+
+func resourceGitlabProjectApprovalRule() *schema.Resource {
+	return &schema.Resource{
+		Schema: resourceGitLabProjectApprovalRuleSchema,
+		Create: resourceGitlabProjectApprovalRuleCreate,
+		Read:   resourceGitlabProjectApprovalRuleRead,
+		Update: resourceGitlabProjectApprovalRuleUpdate,
+		Delete: resourceGitlabProjectApprovalRuleDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+	}
+}
+
+func resourceGitlabProjectApprovalRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	var err error
+
+	options := gitlab.CreateProjectLevelRuleOptions{
+		Name:              gitlab.String(d.Get("name").(string)),
+		ApprovalsRequired: gitlab.Int(d.Get("approvals_required").(int)),
+		UserIDs:           expandApproverIds(d.GetOk("user_ids")),
+		GroupIDs:          expandApproverIds(d.GetOk("group_ids")),
+	}
+	project := d.Get("project").(string)
+
+	log.Printf("[DEBUG] Project %s create gitlab project-level rule %+v", project, options)
+
+	client := meta.(*gitlab.Client)
+	if rule, _, err := client.Projects.CreateProjectApprovalRule(project, &options); err == nil {
+		d.SetId(createApprovalRuleID(rule.ID, project))
+
+		return resourceGitlabProjectApprovalRuleRead(d, meta)
+	}
+
+	return err
+}
+
+func resourceGitlabProjectApprovalRuleRead(d *schema.ResourceData, meta interface{}) error {
+	var err error
+
+	log.Printf("[DEBUG] read gitlab project-level rule %s", d.Id())
+
+	d.Partial(true)
```

Comment:
Is this really needed?

Reply:
Because of the bug described below, Terraform may potentially run into a failure condition while updating the resource. Using partial tells Terraform that everything that's actually been updated so far in the run should be added to the stored state and everything else should not be discarded.